javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala

```diff
@@ -0,0 +1,152 @@
+    val low = f * other.f
+    val high = Math.unsignedMultiplyHigh(f, other.f)
+
+    val nlz = jlLong.numberOfLeadingZeros(high)
+    val excessBits = 64 - nlz
+    val shiftedF =
+      if (nlz == 0) high
+      else (high << nlz) | (low >>> excessBits)
+    val newE = (this.e + other.e + excessBits)
+
+    val remainder = low & (-1L >>> nlz)
+    val halfway = 1L << (excessBits - 1)
+
+    val newF = {
+      if (jlLong.compareUnsigned(remainder, halfway) > 0) // roundup
+        shiftedF + 1
+      else if (jlLong.compareUnsigned(remainder, halfway) < 0)
+        shiftedF
+      else if (shiftedF % 2 == 1) // tie, rounddown is odd
+        shiftedF + 1 // round to even
+      else
+        shiftedF
+    }
+    if (newF == 0) {
+      // denormalized by carry, in case shiftedF = 111..111
+      // (111...111 + 1) >> 1
+      new FloatingPoint(Long.MinValue, newE + 1)
+    } else {
+      new FloatingPoint(newF, newE)
+    }
+  }
+
+  def doesOverflowOnRound(fmt: FloatingPointFormat): Boolean = {
+    val (sig, exponent) = roundNormal(fmt)
+
+    // Remove the leading implicit bit
+    val encodedSig: Long = sig - (1 << (fmt.ExplicitSigBits))
+
+    // Adjust the exponent for exponent bias and mantissa shift
+    val encodedExp: Int = exponent + ((1 << (fmt.ExpBits - 1)) - 1) + fmt.ExplicitSigBits
+
+    encodedExp <= 0 || encodedExp >= (1 << fmt.ExpBits) ||
+        encodedSig <= 0 || encodedSig >= (1 << fmt.ExplicitSigBits)
+  }
+
+  /** Round the 64-bit significand to target#SigBits bits with half-to-even. */
+  private[dec2flt] def roundNormal(target: FloatingPointFormat): (Long, Int) = {
+    val excess = SigBits - target.SigBits // > 0
+    val halfway = 1L << (excess - 1)
+    val shiftedF = f >>> excess
+    val remainder = f & ((1L << excess) - 1) // > 0
+    val normalizedE = e + excess
+
+    val normalizedF = if (remainder > halfway) { // roundup
+      shiftedF + 1
+    } else if (remainder < halfway) { // rounddown
+      shiftedF
+    } else if (shiftedF % 2 == 1) { // tie, rownddown (shiftedF) is odd (1 at LSB)
+      shiftedF + 1 // roundup is even
+    } else { // tie, rounddown is even
+      shiftedF
+    }
+    if (normalizedF > target.MaxSig) {
+      (normalizedF >> 1, normalizedE + 1)
+    } else {
+      (normalizedF, normalizedE)
+    }
+  }
+}
+
+object FloatingPoint {
+  private val MASK: Long = 0xffffffffL
+
+  private val SigBits = 64
+
+  def apply(f: BigInteger): FloatingPoint = {
+    val (newF, newE) = normalize(f, 0)
+    new FloatingPoint(newF, newE)
+  }
+
+  def apply(f: BigInteger, e: Int): FloatingPoint = {
+    val (newF, newE) = normalize(f, e)
+    new FloatingPoint(newF, newE)
+  }
+
+  /** Create a FloatingPoint from normalized f and e. */
+  def normalized(f: Long, e: Int): FloatingPoint = {
+    new FloatingPoint(f, e)
+  }
+
+  /** Normalize the given floating point number z = f * 2^e
+   *
+   *  A floating point number is normalized iff beta^(n-1) <= f < beta^n (beta=2, n=64)
```

Comment:
Consider using the actual character β instead of `beta`. That's easier to understand in formulas.